File: HW2/src/fitness.rs
// Define a trait for fitness functions for reusability
pub trait Fitness {
    fn fitness(&self, member: &[f64]) -> f64;
    fn fitness_bitstring(&self, bitstring: &str) -> f64;
    fn decode_bitstring(&self, bitstring: &str) -> (f64, f64);
}

File: final/fitness.py
from abc import ABC, abstractmethod
from typing import List, Tuple

class Fitness(ABC):
    @abstractmethod
    def fitness(self, member: List[float]) -> float:
        pass

    @abstractmethod
    def fitness_bitstring(self, bitstring: str) -> float:
        pass

    @abstractmethod
    def decode_bitstring(self, bitstring: str) -> Tuple[float, float]:
        pass